teps/0065-retry-failed-tasks-on-demand.md

```diff
@@ -0,0 +1,245 @@
+large amount of data and incur cost. This `retry` mechanism has substantial
+value, where each task of the pipeline incurs a significant computing resources, 
+e.g. `tekton` is used as a backend for ML pipelines.
+
+_Why do we need a new `retry` mechanism when we already support retry in 
+`Pipeline` tasks?_ 
+
+The present `retry` field can only be defined at the time of creation of
+pipeline. This is not suitable for use cases, where a manual intervention
+is necessary to decide whether a rerun is required or not.
+For example, if a service outage is causing a particular task failure, then
+retrying `n` times, won't help, unless we wait for the service to be back
+again and retry. For such manual interventions, we need on-demand `retry`
+mechanism.
+
+Another concocted example, if `Pipeline` were to represent a  CI/CD job, then
+tasks represent test suit, stress test and benchmarks. Now, we need a way to
+know whether a failure was due to some regression, or it is due to flakiness
+of jobs itself or temporary service outage. In this case, simply retrying `n`
+number of times does not seem to help with optimal resource consumption.
+
+### Goals
+
+1. Explore both the merits and demerits in having a new mechanism for on-demand
+   retrying, an _only a failed_ pipeline.
+2. A pipeline may either have failed due to some failures in the tasks or may
+   be user invoked cancel request. Retry only the failed/canceled tasks for a
+   failed `pipelineRun`.
+
+### Non-Goals
+
+1. Retry of successful pipeline runs or anything other than a failed pipeline/task
+   run.
+2. Changing existing retry mechanism.
+3. Manage checkpointing of pipeline state or workspaces, etc. A `pipelineRun`'s
+   state stored in etcd is used as is.
+4. Determine, a failed tasks dependencies i.e. figuring out what
+   all dependent tasks are needed to rerun the failed task.
+   
+### Use Cases (optional)
+
+1. `PipelineRun` can be very resource consuming, and are sometimes susceptible to
+   fail due to transient conditions. For example, due to service outage of a 
+   particular service. In such cases, it is not enough to be retried `n` times,
+   a manual invocation of retry is required.
+   
+2. It will be possible to cancel (e.g. preemption) any running `PipelineRun`, and
+   resume at a later point.
+   
+3. In [Kubeflow pipelines with tekton backend] we are running the pipeline again
+   with a new `pipelineRun`. One of the main problems we see is that some users
+   might use `pipelineRun.uid` and `pipelineRun.name` to distinguish their jobs.
+   So we want a retry feature that can keep these Tekton context variables the
+   same (like `pipelineRun.name` and `pipelineRun.uid`) when users retry the same job.
+
+## Requirements
+
+1. On retry, we would want to reuse the exact same `pipelineRun`, rather than creating
+   a new one and may be deleting the old one. This is because, our users use
+   `pipelineRun.uid` and `pipelineRun.name` to distinguish their jobs.
```

Comment:
I'm not convinced that this is a requirement we can easily satisfy, but I'm open for suggestions about how this might look like. One advantage of this approach, from a metrics point of view, is that we could possibly update the pipelinerun duration metric for include the original duration plus the new one, from the moment it was triggered.